stdlib/array.ncl

```diff
@@ -280,5 +280,50 @@
           l
         else
           (sort cmp (parts.right)) @ [first] @ (sort cmp (parts.wrong)),
+
+    optionals : forall a. Bool -> Array a -> Array a
+      | doc m%"
+        Takes a Bool and an Array and returns the array back if it's true,
+        otherwise, it returns an empty array.
+
+        For example:
+        ```nickel
+          optionals true [1, 2, 3] => [1, 2, 3]
+          optionals false [`a, `b, `c] => []
+        ```
+      "%
+      = fun condition values => if condition then values else [],
```

Comment:
Are we certain we need a standard library function for this? I'm not sure `array.optionals some_bool some_array` is really any more descriptive than just using the `if` statement - I personally think it might be _less_ clear.

Reply:
The idea of this function is really stolen from the nixpkgs libraries. They seem to find value in writing, e.g.
```
buildInputs = [ foo bar ]
  ++ lib.optionals some_condition [ baz another_package ]
  ++ lib.optionals some_other_condition [ yet_another_thing im_running_out_of_names ];
```
Since Nix could just as easily use `if-then-else` as we could, I feel that's empirical evidence that the function is useful. Now, of course we probably would like to structure our dependency specifications differently, so maybe that's a moot point.

Reply:
Fair enough! I'm still not convinced this warrants inclusion in the standard library, but I at least understand the context now.

To expand on my thoughts a little more: in general I'm not sure "does this exist in nixpkgs?" is the best way to gauge whether a function warrants inclusion in the standard library. It's easy enough for users to write their own ports of nixpkgs functions they miss in Nickel (with the caveat that we don't yet have a good package management story). I think the more important question is: is this function likely to be useful to a large number of people writing Nickel? And in this case, I'm not convinced that it is.

Additionally, without looking at the definition of the `optional` function, I'd have a hard time telling you what it does based on the code snippet above, whereas:

```
buildInputs = [ foo bar ]
 ++ (if some_condition then [ baz another_package ] else [])
 ++ (if other_condition then [ yet_another_thing im_running_out_of_names ] else [])
```

is likely to be immediately clear even to newcomers to the language.

Reply:
While I'm not sure it's a full justification, the fact that `if-then-else` introduces one additional level of nesting and the trailing `else []` which is always the same is enough, I fear, to make people write their own version of `lib.optionals` anyway (well, that's what they did in Nix, as it's not in the stdlib). But there may be a case for putting it in, say, the Nixel library, and not in the standard library.

It's also a good point that in Nickel we might use mostly records instead of arrays. I'm fine with not including this right away in the stdlib. I think we've used it here and there but I don't know if there's evidence that this is required in the 1.0 stdlib.